go/store/val/value_store.go

```diff
@@ -16,14 +16,160 @@ package val
 
 import (
 	"context"
-
 	"github.com/dolthub/dolt/go/store/hash"
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/types"
 )
 
+const BytePeekLength = 128
+
 // ValueStore is an interface for a key-value store that can store byte sequences, keyed by a content hash.
 // The only implementation is tree.NodeStore, but ValueStore can be used without depending on the tree package.
 // This is useful for type handlers.
 type ValueStore interface {
 	ReadBytes(ctx context.Context, h hash.Hash) ([]byte, error)
 	WriteBytes(ctx context.Context, val []byte) (hash.Hash, error)
 }
+
+// ImmutableValue represents a content-addressed value stored in a ValueStore.
+// The contents are loaded lazily and stored in |Buf|
+type ImmutableValue struct {
+	Addr hash.Hash
+	Buf  []byte
+	vs   ValueStore
+}
+
+func NewImmutableValue(addr hash.Hash, vs ValueStore) ImmutableValue {
+	return ImmutableValue{Addr: addr, vs: vs}
+}
+
+func (t *ImmutableValue) GetBytes(ctx context.Context) ([]byte, error) {
+	if t.Buf == nil {
+		if t.Addr.IsEmpty() {
+			t.Buf = []byte{}
+			return t.Buf, nil
+		}
+		buf, err := t.vs.ReadBytes(ctx, t.Addr)
+		if err != nil {
+			return nil, err
+		}
+		t.Buf = buf
+	}
+	return t.Buf, nil
+}
+
+type TextStorage struct {
+	ImmutableValue
+	maxByteLength int64
+}
+
+func (t TextStorage) IsExactLength() bool {
```

Comment:
Should comment on all the public methods, naming the interface they implement is fine